File: iOS/FlowWalletKit/Sources/Keys/PrivateKey.swift
/// FlowWalletKit - Private Key Implementation
///
/// This module implements key management using raw private keys.
/// It provides functionality for:
/// - Creating and managing private keys
/// - Secure storage and encryption of private keys
/// - Transaction signing with ECDSA keys
/// - Support for both NIST P-256 and secp256k1 curves
/// - KeyStore JSON import/export

import CryptoKit
import Flow
import Foundation
import KeychainAccess
import WalletCore

/// Implementation of KeyProtocol using raw private keys
public class PrivateKey: KeyProtocol {
    /// Type used for advanced key creation options (raw private key string)
    public typealias Advance = String

    // MARK: - Properties

    /// Storage implementation for persisting key data
    public var storage: any StorageProtocol
    /// Type identifier for this key implementation
    public var keyType: KeyType = .privateKey
    /// Underlying private key implementation from WalletCore
    public let pk: WalletCore.PrivateKey

    // MARK: - Initialization

    /// Initialize with custom storage
    /// - Parameter storage: Storage implementation
    init(storage: any StorageProtocol) {
        pk = WalletCore.PrivateKey()
        self.storage = storage
    }

    /// Initialize with existing private key and storage
    /// - Parameters:
    ///   - pk: WalletCore private key
    ///   - storage: Storage implementation
    init(pk: WalletCore.PrivateKey, storage: any StorageProtocol) {
        self.pk = pk
        self.storage = storage
    }
    
    /// Initialize with existing private key and storage
    /// - Parameters:
    ///   - pk: Private Key raw data
    ///   - storage: Storage implementation
    init?(pk: Data, storage: any StorageProtocol) {
        guard let pk = WalletCore.PrivateKey(data: pk) else {
            return nil
        }
        self.pk = pk
        self.storage = storage
    }

    // MARK: - Key Creation

    /// Create a new private key with storage
    /// - Parameter storage: Storage implementation
    /// - Returns: New private key instance
    public static func create(storage: any StorageProtocol) throws -> PrivateKey {
        let pk = WalletCore.PrivateKey()
        return PrivateKey(pk: pk, storage: storage)
    }

    /// Create and store a new private key
    /// - Parameters:
    ///   - id: Unique identifier for the key
    ///   - password: Password for encrypting the key
    /// - Returns: New private key instance
    /// - Throws: FWKError if encryption or storage fails
    public func create(id: String, password: String) throws -> PrivateKey {
        let pk = WalletCore.PrivateKey()
        guard let cipher = ChaChaPolyCipher(key: password) else {
            throw FWKError.initChaChapolyFailed
        }

        let encrypted = try cipher.encrypt(data: pk.data)
        try storage.set(id, value: encrypted)
        return PrivateKey(pk: pk, storage: storage)
    }

    /// Create and store a new private key (static version)
    /// - Parameters:
    ///   - id: Unique identifier for the key
    ///   - password: Password for encrypting the key
    ///   - storage: Storage implementation
    /// - Returns: New private key instance
    /// - Throws: FWKError if encryption or storage fails
    public static func createAndStore(id: String, password: String, storage: any StorageProtocol) throws -> PrivateKey {
        let pk = WalletCore.PrivateKey()
        guard let cipher = ChaChaPolyCipher(key: password) else {
            throw FWKError.initChaChapolyFailed
        }

        let encrypted = try cipher.encrypt(data: pk.data)
        try storage.set(id, value: encrypted)
        return PrivateKey(pk: pk, storage: storage)
    }

    // MARK: - Key Recovery

    /// Retrieve a stored private key
    /// - Parameters:
    ///   - id: Unique identifier for the key
    ///   - password: Password for decrypting the key
    ///   - storage: Storage implementation
    /// - Returns: Retrieved private key instance
    /// - Throws: FWKError if retrieval or decryption fails
    public static func get(id: String, password: String, storage: any StorageProtocol) throws -> PrivateKey {
        guard let data = try storage.get(id) else {
            throw FWKError.emptyKeychain
        }

        guard let cipher = ChaChaPolyCipher(key: password) else {
            throw FWKError.initChaChapolyFailed
        }

        let pkData = try cipher.decrypt(combinedData: data)

        guard let pk = WalletCore.PrivateKey(data: pkData) else {
            throw FWKError.initPrivateKeyFailed
        }

        return PrivateKey(pk: pk, storage: storage)
    }

    /// Restore a private key from raw data
    /// - Parameters:
    ///   - secret: Raw private key data
    ///   - storage: Storage implementation
    /// - Returns: Restored private key instance
    /// - Throws: FWKError if restoration fails
    public static func restore(secret: Data, storage: any StorageProtocol) throws -> PrivateKey {
        guard let pk = WalletCore.PrivateKey(data: secret) else {
            throw FWKError.restoreWalletFailed
        }
        return PrivateKey(pk: pk, storage: storage)
    }

    /// Restore a private key from KeyStore JSON
    /// - Parameters:
    ///   - json: KeyStore JSON string
    ///   - password: Password to decrypt the KeyStore
    ///   - storage: Storage implementation
    /// - Returns: Restored private key instance
    /// - Throws: FWKError if JSON parsing or decryption fails
    public static func restore(json: String, password: String, storage: any StorageProtocol) throws -> PrivateKey {
        guard let jsonData = json.data(using: .utf8), let passwordData = password.data(using: .utf8) else {
            throw FWKError.restoreWalletFailed
        }

        guard let storedKey = StoredKey.importJSON(json: jsonData) else {
            throw FWKError.invaildKeyStoreJSON
        }
        guard let pkData = storedKey.decryptPrivateKey(password: passwordData) else {
            throw FWKError.invaildKeyStorePassword
        }

        guard let pk = WalletCore.PrivateKey(data: pkData) else {
            throw FWKError.invaildPrivateKey
        }

        return PrivateKey(pk: pk, storage: storage)
    }

    // MARK: - Key Operations

    /// Store the private key securely
    /// - Parameters:
    ///   - id: Unique identifier for the key
    ///   - password: Password for encrypting the key
    /// - Throws: FWKError if encryption or storage fails
    public func store(id: String, password: String) throws {
        guard let cipher = ChaChaPolyCipher(key: password) else {
            throw FWKError.initChaChapolyFailed
        }

        let encrypted = try cipher.encrypt(data: pk.data)
        try storage.set(id, value: encrypted)
    }

    /// Verify a signature
    /// - Parameters:
    ///   - signature: Signature to verify
    ///   - message: Original message that was signed
    ///   - signAlgo: Signature algorithm used
    /// - Returns: Whether the signature is valid
    public func isValidSignature(signature: Data, message: Data, signAlgo: Flow.SignatureAlgorithm) -> Bool {
        guard let pubK = try? getPublicKey(signAlgo: signAlgo) else {
            return false
        }
        return pubK.verify(signature: signature, message: message)
    }

    /// Get the public key for a signature algorithm
    /// - Parameter signAlgo: Signature algorithm
    /// - Returns: Public key data (uncompressed, without prefix)
    public func publicKey(signAlgo: Flow.SignatureAlgorithm) -> Data? {
        guard let pubK = try? getPublicKey(signAlgo: signAlgo) else {
            return nil
        }
        return pubK.uncompressed.data.dropFirst()
    }

    /// Get the raw private key data
    /// - Parameter signAlgo: Signature algorithm (default: ECDSA_P256)
    /// - Returns: Raw private key data
    public func privateKey(signAlgo: Flow.SignatureAlgorithm = .ECDSA_P256) -> Data? {
        return pk.data
    }

    /// Sign data using specified algorithms
    /// - Parameters:
    ///   - data: Data to sign
    ///   - signAlgo: Signature algorithm
    ///   - hashAlgo: Hash algorithm
    /// - Returns: Signature data
    /// - Throws: FWKError if signing fails
    public func sign(data: Data, signAlgo: Flow.SignatureAlgorithm, hashAlgo: Flow.HashAlgorithm) throws -> Data {
        let hashed = try hashAlgo.hash(data: data)
        guard let curve = signAlgo.WCCurve else {
            throw FWKError.unsupportSignatureAlgorithm
        }
        guard let signature = pk.sign(digest: hashed, curve: curve) else {
            throw FWKError.signError
        }

        return signature.dropLast()
    }

    // MARK: - Private Methods

    /// Get the public key implementation for a signature algorithm
    /// - Parameter signAlgo: Signature algorithm
    /// - Returns: Public key implementation
    /// - Throws: FWKError if algorithm is not supported
    private func getPublicKey(signAlgo: Flow.SignatureAlgorithm) throws -> PublicKey {
        switch signAlgo {
        case .ECDSA_P256:
            return pk.getPublicKeyNist256p1()
        case .ECDSA_SECP256k1:
            return pk.getPublicKeySecp256k1(compressed: false)
        case .unknown:
            throw FWKError.unsupportSignatureAlgorithm
        }
    }
}

File: Android/wallet/src/main/java/com/flow/wallet/keys/PrivateKey.kt
package com.flow.wallet.keys

import android.util.Log
import com.flow.wallet.crypto.ChaChaPolyCipher
import com.flow.wallet.crypto.HasherImpl
import com.flow.wallet.errors.WalletError
import com.flow.wallet.storage.StorageProtocol
import org.onflow.flow.models.HashingAlgorithm
import org.onflow.flow.models.SigningAlgorithm
import wallet.core.jni.Curve
import wallet.core.jni.PrivateKey as TWPrivateKey
/**
 * Implementation of KeyProtocol using raw private keys
 * Uses Trust WalletCore for cryptographic operations with enhanced security
 */
class PrivateKey(
    internal var pk: TWPrivateKey,
    override var storage: StorageProtocol,
    private val keyProperties: Map<String, Any> = emptyMap()
) : KeyProtocol, PrivateKeyProvider {
    companion object {
        private const val TAG = "PrivateKey"
        private const val KEY_SIZE = 256
        private const val KEY_ALGORITHM = "EC"

        /**
         * Create a new private key with storage
         */
        fun create(storage: StorageProtocol): PrivateKey {
            val pk = TWPrivateKey()
            return PrivateKey(pk, storage, mapOf(
                "createdAt" to System.currentTimeMillis(),
                "keySize" to KEY_SIZE,
                "algorithm" to KEY_ALGORITHM
            ))
        }

        /**
         * Create and store a new private key
         */
        fun createAndStore(id: String, password: String, storage: StorageProtocol): PrivateKey {
            val pk = TWPrivateKey()
            val cipher = ChaChaPolyCipher(password)
            val encrypted = cipher.encrypt(pk.data())
            storage.set(id, encrypted)
            storage.set("${id}_metadata", mapOf(
                "createdAt" to System.currentTimeMillis(),
                "keySize" to KEY_SIZE,
                "algorithm" to KEY_ALGORITHM
            ).toString().toByteArray())
            return PrivateKey(pk, storage)
        }

        /**
         * Retrieve a stored private key
         */
        fun get(id: String, password: String, storage: StorageProtocol): PrivateKey {
            val encryptedData = storage.get(id) ?: throw WalletError.EmptyKeychain
            val cipher = ChaChaPolyCipher(password)
            val pkData = cipher.decrypt(encryptedData)
            val pk = TWPrivateKey(pkData)
            return PrivateKey(pk, storage, mapOf(
                "retrievedAt" to System.currentTimeMillis()
            ))
        }

        /**
         * Restore a private key from raw data
         */
        fun restore(secret: ByteArray, storage: StorageProtocol): PrivateKey {
            val pk = TWPrivateKey(secret)
            return PrivateKey(pk, storage, mapOf(
                "restoredAt" to System.currentTimeMillis()
            ))
        }
    }

    override val keyType: KeyType = KeyType.PRIVATE_KEY
    override val isHardwareBacked: Boolean = false
    override val advance: Any = Unit

    override val key: Any
        get() = pk

    override val secret: ByteArray
        get() = pk.data()

    override val id: String
        get() = publicKey(SigningAlgorithm.ECDSA_P256)?.let {
            com.google.common.io.BaseEncoding.base64().encode(it)
        } ?: throw WalletError.InitPrivateKeyFailed

    /**
     * Get key properties
     */
    fun getKeyProperties(): Map<String, Any> {
        return buildMap {
            putAll(keyProperties)
            put("keySize", KEY_SIZE)
            put("algorithm", KEY_ALGORITHM)
            put("isHardwareBacked", isHardwareBacked)
        }
    }

    override suspend fun create(advance: Any, storage: StorageProtocol): KeyProtocol {
        return create(storage)
    }

    override suspend fun create(storage: StorageProtocol): KeyProtocol {
        return create(storage)
    }

    override suspend fun createAndStore(id: String, password: String, storage: StorageProtocol): KeyProtocol {
        return createAndStore(id, password, storage)
    }

    override suspend fun get(id: String, password: String, storage: StorageProtocol): KeyProtocol {
        return get(id, password, storage)
    }

    override suspend fun restore(secret: ByteArray, storage: StorageProtocol): KeyProtocol {
        return restore(secret, storage)
    }

    override fun publicKey(signAlgo: SigningAlgorithm): ByteArray? {
        val publicKey: wallet.core.jni.PublicKey?
        try {
            publicKey = when (signAlgo) {
                SigningAlgorithm.ECDSA_P256 -> pk.getPublicKeyNist256p1().uncompressed()
                SigningAlgorithm.ECDSA_secp256k1 -> pk.getPublicKeySecp256k1(false)
                else -> null
            }
            return publicKey?.data()
        } catch (e: Exception) {
            Log.e(TAG, "Failed to get public key", e)
            return null
        } finally {
        }
    }

    override fun privateKey(signAlgo: SigningAlgorithm): ByteArray? {
        return pk.data()
    }

    override suspend fun sign(data: ByteArray, signAlgo: SigningAlgorithm, hashAlgo: HashingAlgorithm): ByteArray {
        return try {
            val hashed = HasherImpl.hash(data, hashAlgo)
            val curve = when (signAlgo) {
                SigningAlgorithm.ECDSA_P256      -> Curve.NIST256P1
                SigningAlgorithm.ECDSA_secp256k1 -> Curve.SECP256K1
                else -> throw WalletError.UnsupportedSignatureAlgorithm
            }

            val fullSignature = pk.sign(hashed, curve)

            // Flow blockchain expects 64-byte signatures for all ECDSA algorithms
            // TrustWallet Core sometimes includes a recovery ID byte, making signatures 65 bytes
            // We need to trim this for both ECDSA_P256 and ECDSA_secp256k1
            val sig = if (fullSignature.size == 65) {
                Log.d(TAG, "Trimming recovery ID from 65-byte signature for $signAlgo")
                fullSignature.copyOfRange(0, 64)
            } else {
                Log.d(TAG, "Using signature as-is (${fullSignature.size} bytes) for $signAlgo")
                fullSignature
            }

            sig
        } catch (e: Exception) {
            Log.e(TAG, "Signing failed", e)
            throw WalletError.SignError
        }
    }

    override fun isValidSignature(signature: ByteArray, message: ByteArray, signAlgo: SigningAlgorithm, hashAlgo: HashingAlgorithm): Boolean {
        val publicKey: wallet.core.jni.PublicKey?
        return try {
            publicKey = when (signAlgo) {
                SigningAlgorithm.ECDSA_P256 -> pk.publicKeyNist256p1.uncompressed()
                SigningAlgorithm.ECDSA_secp256k1 -> pk.getPublicKeySecp256k1(false)
                else -> return false
            }
            val hashed = HasherImpl.hash(message, hashAlgo)
            publicKey.verify(hashed, signature)

        } catch (e: Exception) {
            Log.e(TAG, "Signature verification failed", e)
            false
        } finally {
        }
    }

    override suspend fun store(id: String, password: String) {
        val cipher = ChaChaPolyCipher(password)
        val encrypted = cipher.encrypt(pk.data())
        storage.set(id, encrypted)
        storage.set("${id}_metadata", getKeyProperties().toString().toByteArray())
    }

    override suspend fun remove(id: String) {
        storage.remove(id)
        storage.remove("${id}_metadata")
    }

    override fun allKeys(): List<String> {
        return storage.allKeys
    }

    /**
     * Export private key in specified format
     * @param format Export format
     * @return Exported key data
     */
    override fun exportPrivateKey(format: KeyFormat): ByteArray {
        return when (format) {
            KeyFormat.PKCS8 -> pk.data()
            KeyFormat.RAW -> pk.data()
            else -> throw WalletError.UnsupportedKeyFormat
        }
    }

    /**
     * Import private key from data
     * @param data Key data
     * @param format Import format
     */
    override fun importPrivateKey(data: ByteArray, format: KeyFormat) {
        when (format) {
            KeyFormat.PKCS8, KeyFormat.RAW -> {
                if (data.isEmpty()) {
                    throw WalletError.EmptyKey
                }
                try {
                    pk = TWPrivateKey(data)
                } catch (e: Exception) {
                    Log.e(TAG, "Failed to import private key", e)
                    throw WalletError.InvalidPrivateKey
                }
            }
            else -> throw WalletError.UnsupportedKeyFormat
        }
    }

    /**
     * Clean up sensitive data
     */
    fun cleanup() {
        pk = TWPrivateKey()
    }
} 